OSDOCS#4549: Adding docs for exposing services on firewalls (#65738)
openshift/openshift-docs: modules/microshift-firewall-update-for-service.adoc

// Module included in the following assemblies:
//
// * microshift_networking/microshift-firewall.adoc

:_content-type: CONCEPT
[id="microshift-firewall-update-for-service_{context}"]
= Updating firewall ports when a service is exposed

Firewalld is often active when you run services on a {product-title}. This can disrupt certain services on a {product-title} because traffic to the ports might be blocked by the firewall. You need to ensure the necessary firewall ports are open if you want certain services to be accessible from outside the host. There are several options for opening your ports:

* Services of the `NodePort` and `LoadBalancer` type are automatically available with OVN-K.
+

> In these cases, OVN-K adds iptable rules so the traffic to the cluster IP address is delivered to the relevant ports. This is done using the PREROUTING rule chain and is then forwarded to the OVN-K to bypass the firewalld rules for local host ports and services.

Reviewer: Maybe we shall add some background knowledge like: 
Both iptables and firewalld are backed by nftables in RHEL9. To avoid conflict, the nftables rules generated by iptables always have precedence over the ones generated by firewalld.

> In these cases, OVN-K adds iptables rules so the traffic to the cluster IP address is delivered to the relevant ports. This is done using the PREROUTING rule chain and is then forwarded to the OVN-K to bypass the firewalld rules for local host ports and services. Iptables and firewalld are backed by nftables in RHEL9. The nftables rules, that the iptables generates, always has priority over the rules that the firewalld generates.

Reviewer: `the traffic to the cluster IP address` should be `the traffic to the node IP address`

* Pods with the `HostPort` parameter settings are automatically available. This also includes the router-default pod, which uses ports 80 and 443.
+

> For `HostPort` pods, the CRI-O config sets up DNAT to the pods IP address and port.

Reviewer: CRI-O does this with iptables rules as well.

> For `HostPort` pods, the CRI-O config sets up DNAT to the pods IP address and port. The CRI-O config sets up the pods with the iptables rules as well.

Reviewer: For `HostPort` pods, the CRI-O config sets up DNAT iptables rules to the pods IP address and port.

Author: Oh gotcha, my misunderstanding here, thank you!

// Module included in the following assemblies:
//
// * microshift_networking/microshift-firewall.adoc

:_content-type: CONCEPT
[id="microshift-firewall-update-for-service_{context}"]

> = Updating firewall ports when a service is exposed

Reviewer 2: If this is a concept module, then it shouldn't be a gerund verb. Suggest rewriting to be more of a noun statement

> Firewalld is often active when you run services on {microshift-short}. This can disrupt certain services on {microshift-short} because traffic to the ports might be blocked by the firewall. You need to ensure the necessary firewall ports are open if you want certain services to be accessible from outside the host. There are several options for opening your ports:

Reviewer 2: ```suggestion
Firewalld is often active when you run services on {microshift-short}. This can disrupt certain services on {microshift-short} because traffic to the ports might be blocked by the firewall. You must ensure that the necessary firewall ports are open if you want certain services to be accessible from outside the host. There are several options for opening your ports:
```

> Firewalld is often active when you run services on a {microshift-short}. This can disrupt certain services on a {microshift-short} because traffic to the ports might be blocked by the firewall. You need to ensure the necessary firewall ports are open if you want certain services to be accessible from outside the host. There are several options for opening your ports:

Author: Nice catch! Meant to update that when I added the new {microshfit-short} attribute.

> * Services of the `NodePort` and `LoadBalancer` type are automatically available with OVN-K.

Reviewer 2: I'm having trouble figuring out what the two options are with how these are written. I would try to make it clearer.

The first one talks about services of these two types. What about of the other types? And are they only available with OVN-K? (like, is there an option other than OVN-K that users could be using where these wouldn't be availabel automatically?

Author: I think they mainly just wanted to reference the two types. But I'm also pretty sure that OVN-K is automatically part of MicroShift the assumption is that the users use the services with OVN-K. 
Lemme ask about that though, becuase my experience with firewall ports is limited

+

> In these cases, OVN-K adds iptables rules so the traffic to the node IP address is delivered to the relevant ports. This is done using the PREROUTING rule chain and is then forwarded to the OVN-K to bypass the firewalld rules for local host ports and services. Iptables and firewalld are backed by nftables in RHEL9. The nftables rules, that the iptables generates, always has priority over the rules that the firewalld generates.

Reviewer 2: * Is "OVN-K" the official name, or should it be "OVN-Kubernetes"?
* Should write out "RHEL9". We have attributes in the common-attributes file for long and short names
* "that the iptables generates" -> "which the iptables generates"

Author: Its meant to be OVN-Kubernetes, I just found out that is the right format, also need to update that attribute

> * Pods with the `HostPort` parameter settings are automatically available. This also includes the router-default pod, which uses ports 80 and 443.

Reviewer 2: * Same confusion here. What about ports that don't have this parameter set?
* Should `router-default` be in backticks?

Author: I still think this concept is meant to be for certain cases, but will also check

+

> For `HostPort` pods, the CRI-O config sets up iptables DNAT rules to the pods IP address and port.

Reviewer 2: Should "pods" be "pod's"?

> These methods function for clients on the same host or remote host. The iptables rules, that are added by OVN-K and CRI-O, attach to the PREROUTING and OUTPUT chains. The local traffic goes through the OUTPUT chain with the interface set to the `lo` type. The DNAT runs before it hits filler rules in the INPUT chain.

Reviewer 2: * Should some of these items be in backticks?
* Same question about "OVN-K"
* Should people know what DNAT is, or does it need spelled out on first use above?
* A few wording suggestions, as long as it doesn't change the meaning

```suggestion
These methods function for clients whether they are on the same host or on a remote host. The iptables rules, which are added by OVN-K and CRI-O, attach to the `PREROUTING` and `OUTPUT` chains. The local traffic goes through the `OUTPUT` chain with the interface set to the `lo` type. The DNAT runs before it hits filler rules in the `INPUT` chain.
```

> The {microshift-short} API server does not run in CRI-O, therefore the API server is subject to the firewall configurations. You can open port 6443 in the firewall to access the API server in your {microshift-short}. 

Reviewer 2: * Seems like this is missing "cluster" at the end here.
* Optional suggested reword

```suggestion
Because the {microshift-short} API server does not run in CRI-O, it is subject to the firewall configurations. You can open port 6443 in the firewall to access the API server in your {microshift-short} cluster. 
```
